doc/ir-semantics.text

```diff
@@ -58,30 +73,71 @@ Evaluation of simple expressions:
   eval H E primop(x₁, ..., xₙ) = 〚primop〛((H,E)[x₁], ..., (H,E)[xₙ])
 
 Reduction relation
-  (T, H, E, P, l) -> ((T, (H,E)[x]), H, E, P, l + 1)
+  (T, H, E, C, P, l) -> ((T, (H,E)[x]), H, E, C, P, l + 1)
     when P(l) = print x
 
-  (T, H, E, P, l) -> (T, H, E, P, l')
+  (T, H, E, C, P, l) -> (T, H, E, C, P, l')
     where P(l) = branch x l₁ l₂
       and l' := if (H,E)[x] then l₁ else l₂
 
-  (T, H, E, P, l) -> (T, H, (E, x ↦ v), P, l+1)
+  (T, H, E, C, P, l) -> (T, H, (E, x ↦ v), C, P, succ l)
     when P(l) = (const x = e)
      and v := eval H E x
 
-  (T, H, E, P, l) -> (T, (H, a ↦ v), (E, x ↦ a), P, l+1)
+  (T, H, E, C, P, l) -> (T, (H, a ↦ v), (E, x ↦ a), C, P, succ l)
     when P(l) = (mut x = e)
      and v := eval H E e
      and a fresh
 
-  (T, H, E, P, l) -> (T, (H,E)[x] := v, E, P, l+1)
+  (T, H, E, C, P, l) -> (T, (H,E)[x] := v, E, C, P, succ l)
     when P(l) = (x ← e)
      and v := eval H E x
 
-  (T, H, E, P, l) -> (T, H, E, P, l+1)
-    when P(l) = invalidate(e, l', x̄)
+  (T, H, E, C, P, l) -> (T, H, E, C, P, succ l)
+    when P(l) = osr(e, _, _)
+     and (eval H E e) is not (true)
+
+  (T, H, E, C, P, l) -> (T, H, E', C, P', l')
+    when P(l) = osr(e, v, l', (x = ee)*)
      and (eval H E e) is (true)
+     and P' := C(v)
+     and E' := (x ↦ ⊥ | (eval H E ee))*
```

Comment:
since eval is not defined for ⊥ I wanted to say that e.g. (x = ⊥, y = 1+2) maps to the environment {x -> ⊥, y -> eval 1+2}

but yeah the notation could be improved.

adding the mut/const modifiers sounds like a good idea, though note that the target scope should fully define whether they are const or mut